### PR TITLE
feat(protocol-designer, step-generation): custom z offset for blowout

### PR DESCRIPTION
**Pull request by jerader**
closes AUTH-7

# Overview

This PR adds a custom z offset field to blowout in the advanced settings of mix and transfer

# Test Plan

create a flex or Ot-2 protocol and add a transfer step. open the advanced settings and click on the blowout checkbox. The input field for z offset should be next to the location drop down. It should be disabled if the location is not chosen, or if the source/dest well is a trash bin/waste chute. Once it is enabled, click on it and it should render the tip position modal. The default should be 0 at the top of the well and then the range should be the depth of the well in a negative number. Add a custom offset and the step should be created successfully.

Now, do the same thing with a mix step.

# Changelog

- create a new field for blowout z offset
- add new type to fields and step-generation
- modify the tip position modal to account for the z offset starting from the top
- fix the tests
- fix cypress, the blowout commands now generate offset from the top of the well instead of from the bottom (this theoretically should NOT be a change in behavior since before, we were getting the z offset that was the height of the well from the bottom)

# Review requests

see test plan

# Risk assessment

low

### CODE REVIEW

File: step-generation/src/commandCreators/atomic/blowout.ts [hunk 1]
```diff
@@ -1,8 +1,7 @@
 import { uuid, getLabwareSlot } from '../../utils'
 import { COLUMN_4_SLOTS } from '../../constants'
 import * as errorCreators from '../../errorCreators'
-import type { CreateCommand } from '@opentrons/shared-data'
-import type { BlowoutParams } from '@opentrons/shared-data/protocol/types/schemaV3'
```
Comment (jerader) [inline step-generation/src/commandCreators/atomic/blowout.ts:5]: finally migrated to using the latest schema for blowout props 😃

File: protocol-designer/src/components/StepEditForm/fields/BlowoutZOffsetField.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+type BlowoutZOffsetFieldProps = FieldProps & {
+  destLabwareId: unknown
+  sourceLabwareId?: unknown
+  blowoutLabwareId?: unknown
+}
+
+export const BlowoutZOffsetField = (
+  props: BlowoutZOffsetFieldProps
+): JSX.Element => {
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/BlowoutZOffsetField.tsx]: for consistency

```suggestion
export function BlowoutZOffsetField(
  props: BlowoutZOffsetFieldProps
): JSX.Element {
```

File: protocol-designer/src/components/StepEditForm/fields/BlowoutZOffsetField.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import { ZTipPositionModal } from './TipPositionField/ZTipPositionModal'
+import type { FieldProps } from '../types'
+
+import styles from '../StepEditForm.module.css'
+
+type BlowoutZOffsetFieldProps = FieldProps & {
+  destLabwareId: unknown
+  sourceLabwareId?: unknown
+  blowoutLabwareId?: unknown
+}
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/BlowoutZOffsetField.tsx]: i think basically we use interface so if there is no specific reason, using interface will be good.

```ts
interface BlowoutZOffsetFieldProps extends FieldProps {
  destLabwareId: unknown;
  sourceLabwareId?: unknown;
  blowoutLabwareId?: unknown;
}
```

File: protocol-designer/src/components/StepEditForm/fields/BlowoutZOffsetField.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          wellDepthMm={labwareZDimension}
+        />
+      ) : null}
+      <div {...targetProps}>
+        <InputField
+          data-testid="BlowoutZOffsetField_inputField"
+          disabled={disabled}
+          className={styles.small_field}
+          readOnly
+          onClick={disabled ? () => {} : () => setModalOpen(true)}
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/BlowoutZOffsetField.tsx]: nit

```suggestion
          onClick={disabled ? undefined : () => setModalOpen(true)}
```

File: protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionZAxisViz.tsx [hunk 5]
```diff
@@ -8,19 +8,28 @@ import styles from './TipPositionInput.module.css'
...
   wellDepthMm: number
+  mmFromBottom?: number
+  mmFromTop?: number
 }
 
-export const TipPositionZAxisViz = (props: Props): JSX.Element => {
-  const fractionOfWellHeight = props.mmFromBottom / props.wellDepthMm
+export const TipPositionZAxisViz = (
+  props: TipPositionZAxisVizProps
+): JSX.Element => {
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionZAxisViz.tsx]:
```suggestion
export function TipPositionZAxisViz (
  props: TipPositionZAxisVizProps
): JSX.Element {
```

File: protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionZAxisViz.tsx [hunk 6]
```diff
@@ -8,19 +8,28 @@ import styles from './TipPositionInput.module.css'
...
+  const { mmFromBottom, mmFromTop, wellDepthMm } = props
+  let positionInTube = 0
+  if (mmFromBottom != null) {
+    positionInTube = mmFromBottom
+  } else if (mmFromTop != null) {
+    positionInTube = mmFromTop
+  }
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionZAxisViz.tsx]: nit

```suggestion
const positionInTube = mmFromBottom ?? mmFromTop ?? 0
```

File: protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionZAxisViz.tsx [hunk 7]
```diff
@@ -8,19 +8,28 @@ import styles from './TipPositionInput.module.css'
...
+  const fractionOfWellHeight = positionInTube / wellDepthMm
   const pixelsFromBottom =
-    Number(fractionOfWellHeight) * WELL_HEIGHT_PIXELS - WELL_HEIGHT_PIXELS
-  const roundedPixelsFromBottom = round(pixelsFromBottom, PIXEL_DECIMALS)
-  const bottomPx = props.wellDepthMm
-    ? roundedPixelsFromBottom
-    : props.mmFromBottom - WELL_HEIGHT_PIXELS
+    Number(fractionOfWellHeight) * WELL_HEIGHT_PIXELS -
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionZAxisViz.tsx]: question
need to use `Number` for `fractionOfWellHeight`?

File: protocol-designer/src/components/StepEditForm/fields/TipPositionField/ZTipPositionModal.tsx [hunk 8]
```diff
@@ -36,21 +37,25 @@ export function ZTipPositionModal(props: ZTipPositionModalProps): JSX.Element {
...
+  const defaultMm = isBlowout
+    ? 0
+    : utils.getDefaultMmFromBottom({
+        name,
+        wellDepthMm,
+      })
 
   const [value, setValue] = React.useState<string | null>(
-    mmFromBottom === null ? null : String(mmFromBottom)
+    zValue === null ? null : String(zValue)
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/TipPositionField/ZTipPositionModal.tsx]: maybe

```suggestion
    zValue !== null ? String(zValue) : zValue
    
    // or 
    
    zValue === null ? zValue : String(zValue)
```

File: protocol-designer/src/components/StepEditForm/fields/TipPositionField/ZTipPositionModal.tsx [hunk 9]
```diff
@@ -36,21 +37,25 @@ export function ZTipPositionModal(props: ZTipPositionModalProps): JSX.Element {
...
   )
   const [isDefault, setIsDefault] = React.useState<boolean>(
-    !isIndeterminate && mmFromBottom === null
+    !isIndeterminate && (isBlowout ? zValue === 0 : zValue === null)
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/TipPositionField/ZTipPositionModal.tsx]: If you simplify this, that will be great.

`isBlowout ? zValue === 0 : zValue === null`

File: protocol-designer/src/components/StepEditForm/fields/__tests__/BlowoutZOffsetField.test.tsx [hunk 10]
```diff
@@ -0,0 +1,57 @@
+import * as React from 'react'
+import { fireEvent, screen } from '@testing-library/react'
+import { fixture96Plate } from '@opentrons/shared-data'
+import { SOURCE_WELL_BLOWOUT_DESTINATION } from '@opentrons/step-generation'
+import { describe, it, vi, beforeEach } from 'vitest'
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/__tests__/BlowoutZOffsetField.test.tsx]:
```suggestion
import { describe, it, vi, beforeEach } from 'vitest'
import { fixture96Plate } from '@opentrons/shared-data'
import { SOURCE_WELL_BLOWOUT_DESTINATION } from '@opentrons/step-generation'
```

File: protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionZAxisViz.tsx (hunk 6 above)
Comment (jerader) [reply protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionZAxisViz.tsx]: wow for some reason i didn't know you could have a nullish coalescing operator twice in a single expression like that

**Comment by koji** [comment]
Screenshot 2024-04-04 at 9 22 20 AM

_well_height_label wouldn't need `right: 10px`

Screenshot 2024-04-04 at 9 24 50 AM

**Review by koji — APPROVED** [review]
sandbox worked as expected.
I just left one comment on modal's well height label position otherwise the changes look good to me.

**Comment by jerader** [comment]
> _well_height_label wouldn't need `right: 10px`
> 

nice! thanks for catching this @koji. i'll remove it. seems like its some old css styling added 6 years ago 😅

**Pull request merged**